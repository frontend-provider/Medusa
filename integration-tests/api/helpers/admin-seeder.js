const Scrypt = require("scrypt-kdf")
const { User } = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const buf = await Scrypt.kdf("secret_password", { logN: 1, r: 1, p: 1 })
  const password_hash = buf.toString("base64")

  await manager.insert(User, {
    id: "admin_user",
    email: "admin@medusa.js",
    api_token: "test_token",
    role: "admin",
    password_hash,
    ...data,
  })
}
