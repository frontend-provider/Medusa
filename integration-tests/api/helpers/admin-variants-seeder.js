const {
  Region,
  Product,
  ProductVariant,
  PriceList,
  CustomerGroup,
  Customer,
  Image,
  ShippingProfile,
  ProductCollection,
  ProductOption,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const yesterday = ((today) => new Date(today.setDate(today.getDate() - 1)))(
    new Date()
  )

  const tenDaysAgo = ((today) => new Date(today.setDate(today.getDate() - 10)))(
    new Date()
  )

  const defaultProfile = await manager.findOne(ShippingProfile, {
    where: { type: ShippingProfile.default },
  })

  const collection = manager.create(ProductCollection, {
    id: "test-collection",
    handle: "test-collection",
    title: "Test collection",
  })

  await manager.save(collection)

  await manager.insert(Region, {
    id: "reg-europe",
    name: "Test Region Europe",
    currency_code: "eur",
    tax_rate: 0,
  })

  await manager.insert(Region, {
    id: "reg-us",
    name: "Test Region US",
    currency_code: "usd",
    tax_rate: 0,
  })

  const customer = await manager.create(Customer, {
    id: "test-customer",
    email: "john@doe.com",
    first_name: "John",
    last_name: "Doe",
    password_hash:
      "c2NyeXB0AAEAAAABAAAAAVMdaddoGjwU1TafDLLlBKnOTQga7P2dbrfgf3fB+rCD/cJOMuGzAvRdKutbYkVpuJWTU39P7OpuWNkUVoEETOVLMJafbI8qs8Qx/7jMQXkN", // password matching "test"
    has_account: true,
  })

  const customerGroup = await manager.create(CustomerGroup, {
    id: "test-group",
    name: "test-group",
  })

  await manager.save(customerGroup)
  customer.groups = [customerGroup]
  await manager.save(customer)

  const priceListActive = await manager.create(PriceList, {
    id: "pl",
    name: "VIP sale",
    description: "All year sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  await manager.save(priceListActive)

  const priceListExpired = await manager.create(PriceList, {
    id: "pl_expired",
    name: "VIP summer sale",
    description: "Summer sale for VIP customers.",
    type: "sale",
    status: "active",
    starts_at: tenDaysAgo,
    ends_at: yesterday,
  })

  await manager.save(priceListExpired)

  const priceListWithCustomers = await manager.create(PriceList, {
    id: "pl_with_customers",
    name: "VIP winter sale",
    description: "Winter sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  priceListWithCustomers.customer_groups = [customerGroup]

  await manager.save(priceListWithCustomers)

  const productMultiReg = manager.create(Product, {
    id: "test-product",
    handle: "test-product-reg",
    title: "Multi Reg Test product",
    profile_id: defaultProfile.id,
    description: "test-product-description",
    status: "published",
    collection_id: "test-collection",
  })

  const image = manager.create(Image, {
    id: "test-image",
    url: "test-image.png",
  })

  productMultiReg.images = [image]

  await manager.save(productMultiReg)

  await manager.save(ProductOption, {
    id: "test-option",
    title: "test-option",
    product_id: "test-product",
  })

  const variantMultiReg = await manager.create(ProductVariant, {
    id: "test-variant",
    inventory_quantity: 10,
    title: "Test variant",
    variant_rank: 0,
    sku: "test-sku",
    ean: "test-ean",
    upc: "test-upc",
    barcode: "test-barcode",
    product_id: "test-product",
    prices: [
      {
        id: "test-price-multi-usd",
        currency_code: "usd",
        type: "default",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-usd",
        currency_code: "usd",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-expired-multi-usd",
        currency_code: "usd",
        amount: 70,
        price_list_id: "pl_expired",
      },
      {
        id: "test-price-multi-eur",
        currency_code: "eur",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-eur",
        currency_code: "eur",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-multi-eur-with-customer",
        currency_code: "eur",
        amount: 40,
        price_list_id: "pl_with_customers",
      },
      {
        id: "test-price-discount-expired-multi-eur",
        currency_code: "eur",
        amount: 70,
        price_list_id: "pl_expired",
      },
    ],
    options: [
      {
        id: "test-variant-option-reg",
        value: "Default variant",
        option_id: "test-option",
      },
    ],
  })

  await manager.save(variantMultiReg)
}
