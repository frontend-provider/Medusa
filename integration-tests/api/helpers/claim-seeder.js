const {
  ClaimOrder,
  Order,
  LineItem,
  Fulfillment,
  Return,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  let orderWithClaim = manager.create(Order, {
    id: "order-with-claim",
    customer_id: "test-customer",
    email: "test@email.com",
    payment_status: "captured",
    fulfillment_status: "fulfilled",
    billing_address: {
      id: "test-billing-address",
      first_name: "lebron",
    },
    shipping_address: {
      id: "test-shipping-address",
      first_name: "lebron",
      country_code: "us",
    },
    region_id: "test-region",
    currency_code: "usd",
    tax_rate: 0,
    discounts: [],
    payments: [
      {
        id: "test-payment-for-claim-order",
        amount: 10000,
        currency_code: "usd",
        amount_refunded: 0,
        provider_id: "test-pay",
        data: {},
      },
    ],
    items: [],
    ...data,
  })

  await manager.save(orderWithClaim)

  const li = manager.create(LineItem, {
    id: "test-item-co-2",
    fulfilled_quantity: 1,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 1,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li)

  const li2 = manager.create(LineItem, {
    id: "test-item-co-3",
    fulfilled_quantity: 4,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 4,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li2)

  const claimWithFulfillment = manager.create(ClaimOrder, {
    id: "claim-w-f",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  const ful1 = manager.create(Fulfillment, {
    id: "fulfillment-co-1",
    data: {},
    provider_id: "test-ful",
  })

  const ful2 = manager.create(Fulfillment, {
    id: "fulfillment-co-2",
    data: {},
    provider_id: "test-ful",
  })

  claimWithFulfillment.fulfillments = [ful1, ful2]

  await manager.save(claimWithFulfillment)

  const claimWithReturn = manager.create(ClaimOrder, {
    id: "claim-w-r",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  await manager.save(claimWithReturn)

  await manager.insert(Return, {
    id: "return-id-2",
    claim_order_id: "claim-w-r",
    status: "requested",
    refund_amount: 0,
    data: {},
  })
}
