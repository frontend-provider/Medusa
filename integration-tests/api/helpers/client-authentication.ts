const AUTH_COOKIE = {}
export async function getClientAuthenticationCookie(
  api,
  email = null,
  password = null
) {
  const user = {
    email: email ?? "test@medusajs.com",
    password: password ?? "test",
  }

  if (AUTH_COOKIE[user.email]) {
    return AUTH_COOKIE[user.email]
  }

  const authResponse = await api.post("/store/auth", user)
  AUTH_COOKIE[user.email] = authResponse.headers["set-cookie"][0].split(";")

  return AUTH_COOKIE[user.email]
}
