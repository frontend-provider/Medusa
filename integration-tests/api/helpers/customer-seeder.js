const { Customer, Address } = require("@medusajs/medusa")

module.exports = async (connection, data = {}) => {
  const manager = connection.manager

  await manager.insert(Customer, {
    id: "test-customer-1",
    email: "test1@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-2",
    email: "test2@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-3",
    email: "test3@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-has_account",
    email: "test4@email.com",
    has_account: true,
  })

  await manager.insert(Address, {
    id: "test-address",
    first_name: "Lebron",
    last_name: "James",
    customer_id: "test-customer-1",
  })
}
