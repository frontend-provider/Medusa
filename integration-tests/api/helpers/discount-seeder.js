const {
  ShippingProfile,
  Region,
  Discount,
  DiscountRule,
} = require("@medusajs/medusa")
module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  await manager.insert(Region, {
    id: "test-region",
    name: "Test Region",
    currency_code: "usd",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })

  await manager.insert(Region, {
    id: "test-region-2",
    name: "Test Region 2",
    currency_code: "eur",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })
}
