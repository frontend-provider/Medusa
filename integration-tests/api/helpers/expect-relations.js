const { expect } = require("@jest/globals")

export function expectRelations(expected, actual){
    expected.forEach(a => expectRelations_(a, actual))
}

function expectRelations_(expected, actual){
    const items = expected.split(".");
    let data_ = actual;

    for(const i in items){
        data_ = data_[items[i]];
        
        if(data_ instanceof Array && data_.length > 0) {
            data_ = data_[0];
        } 
    
        expect(data_).toBeDefined();
    }
}
