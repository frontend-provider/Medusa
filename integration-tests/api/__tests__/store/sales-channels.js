const path = require("path")

const { useApi } = require("../../../helpers/use-api")
const { useDb } = require("../../../helpers/use-db")

const adminSeeder = require("../../helpers/admin-seeder")
const {
  simpleSalesChannelFactory,
  simpleCartFactory, simpleRegionFactory, simpleProductFactory,
} = require("../../factories")

const startServerWithEnvironment =
  require("../../../helpers/start-server-with-environment").default

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

jest.setTimeout(50000)

describe("sales channels", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    const [process, connection] = await startServerWithEnvironment({
      cwd,
      env: { MEDUSA_FF_SALES_CHANNELS: true },
      verbose: false,
    })
    dbConnection = connection
    medusaProcess = process
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("POST /store/cart/", () => {
    let salesChannel
    let disabledSalesChannel

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        await simpleRegionFactory(dbConnection, {
          name: "Test region",
          tax_rate: 0,
        })
        await simpleSalesChannelFactory(dbConnection, {
          name: "Default Sales Channel",
          description: "Created by Medusa",
          is_default: true
        })
        disabledSalesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "disabled cart sales channel",
          description: "disabled cart sales channel description",
          is_disabled: true,
        })
        salesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "cart sales channel",
          description: "cart sales channel description",
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("returns a cart with the default sales channel", async () => {
      const api = useApi()

      const response = await api.post(`/store/carts`, {}, adminReqConfig)

      expect(response.data.cart.sales_channel).toBeTruthy()
      expect(response.data.cart.sales_channel).toEqual(
        expect.objectContaining({
          name: "Default Sales Channel",
          description: "Created by Medusa",
        })
      )
    })

    it("returns a cart with the given sales channel", async () => {
      const api = useApi()

      const response = await api.post(`/store/carts`, { sales_channel_id: salesChannel.id }, adminReqConfig)

      expect(response.data.cart.sales_channel).toBeTruthy()
      expect(response.data.cart.sales_channel).toEqual(
        expect.objectContaining({
          name: salesChannel.name,
          description: salesChannel.description,
        })
      )
    })

    it("throw if the given sales channel is disabled", async () => {
      const api = useApi()

      const err = await api.post(
        `/store/carts`,
        { sales_channel_id: disabledSalesChannel.id },
        adminReqConfig
      ).catch(err => err)

      expect(err.response.status).toEqual(400)
      expect(err.response.data.message).toBe(`Unable to assign the cart to a disabled Sales Channel "disabled cart sales channel"`)
    })
  })

  describe("POST /store/cart/:id", () => {
    let salesChannel1, salesChannel2, disabledSalesChannel
    let product1, product2
    let cart

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        await simpleRegionFactory(dbConnection, {
          name: "Test region",
          currency_code: "usd",
          tax_rate: 0,
        })

        salesChannel1 = await simpleSalesChannelFactory(dbConnection, {
          name: "salesChannel1",
          description: "salesChannel1",
        })
        salesChannel2 = await simpleSalesChannelFactory(dbConnection, {
          name: "salesChannel2",
          description: "salesChannel2",
        })
        disabledSalesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "disabled cart sales channel",
          description: "disabled cart sales channel description",
          is_disabled: true,
        })

        product1 = await simpleProductFactory(
          dbConnection,
          {
            title: "prod 1",
            sales_channels: [salesChannel1],
            variants: [
              {
                id: "test-variant",
                prices: [
                  {
                    amount: 50,
                    currency: "usd",
                    variant_id: "test-variant",
                  },
                ],
              },
            ],
          },
        )
        product2 = await simpleProductFactory(
          dbConnection,
          {
            sales_channels: [salesChannel2],
            variants: [
              {
                id: "test-variant-2",
                prices: [
                  {
                    amount: 100,
                    currency: "usd",
                    variant_id: "test-variant-2",
                  },
                ],
              },
            ],
          },
        )

        cart = await simpleCartFactory(
          dbConnection,
          {
            sales_channel: salesChannel1,
            line_items: [
              {
                variant_id: "test-variant",
                unit_price: 50,
              },
            ],
          },
        )
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it(
        "updates a cart sales channels should remove the items that does not belongs to the new sales channel",
        async () => {
        const api = useApi()

        let response = await api.get(`/store/carts/${cart.id}`, adminReqConfig)

        expect(response.data.cart.sales_channel).toBeTruthy()
        expect(response.data.cart.sales_channel).toEqual(
          expect.objectContaining({
            name: salesChannel1.name,
            description: salesChannel1.description,
          })
        )
        expect(response.data.cart.items.length).toBe(1)
        expect(response.data.cart.items[0].variant.product).toEqual(
          expect.objectContaining({
            id: product1.id,
            title: product1.title,
          })
        )

        response = await api.post(`/store/carts/${cart.id}`, { sales_channel_id: salesChannel2.id }, adminReqConfig)

        expect(response.data.cart.sales_channel).toBeTruthy()
        expect(response.data.cart.sales_channel).toEqual(
          expect.objectContaining({
            name: salesChannel2.name,
            description: salesChannel2.description,
          })
        )
        expect(response.data.cart.items.length).toBe(0)
      }
    )

    it("throw if the given sales channel is disabled", async () => {
      const api = useApi()

      const err = await api.post(
        `/store/carts/${cart.id}`,
        { sales_channel_id: disabledSalesChannel.id },
        adminReqConfig
      ).catch(err => err)

      expect(err.response.status).toEqual(400)
      expect(err.response.data.message).toBe("Unable to assign the cart to a disabled Sales Channel \"disabled cart sales channel\"")
    })
  })

  describe("GET /store/cart/:id", () => {
    let cart

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)

        cart = await simpleCartFactory(dbConnection, {
          sales_channel: {
            name: "test name",
            description: "test description",
          },
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("returns cart with sales channel for single cart", async () => {
      const api = useApi()

      const response = await api.get(`/store/carts/${cart.id}`, adminReqConfig)

      expect(response.data.cart.sales_channel).toBeTruthy()
      expect(response.data.cart.sales_channel).toMatchSnapshot({
        id: expect.any(String),
        name: "test name",
        description: "test description",
        is_disabled: false,
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })
})
