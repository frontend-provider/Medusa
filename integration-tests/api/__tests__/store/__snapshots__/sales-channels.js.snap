// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sales channels GET /store/cart/:id returns cart with sales channel for single cart 1`] = `
Object {
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "test description",
  "id": Any<String>,
  "is_disabled": false,
  "name": "test name",
  "updated_at": Any<String>,
}
`;
